src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```diff
@@ -0,0 +1,687 @@
+	mag_comp.offset_coef[2] = fix_sign(off_z_lsb_msb, 12);
+	mag_comp.offset_coef_t = fix_sign(off_t_lsb, 8) / 5.0f;
+
+	sens_x = otp_data[BMM350_MAG_SENS_X] >> 8;
+	sens_y = (uint8_t)otp_data[BMM350_MAG_SENS_Y];
+	sens_z = otp_data[BMM350_MAG_SENS_Z] >> 8;
+	sens_t = otp_data[BMM350_TEMP_OFF_SENS] >> 8;
+
+	mag_comp.sensit_coef[0] = fix_sign(sens_x, 8) / 256.0f;
+	mag_comp.sensit_coef[1] = (fix_sign(sens_y, 8) / 256.0f) + BMM350_SENS_CORR_Y;
+	mag_comp.sensit_coef[2] = fix_sign(sens_z, 8) / 256.0f;
+	mag_comp.sensit_coef_t = fix_sign(sens_t, 8) / 512.0f;
+
+	tco_x = (otp_data[BMM350_MAG_TCO_X] & BMM350_LSB_MASK);
+	tco_y = (otp_data[BMM350_MAG_TCO_Y] & BMM350_LSB_MASK);
+	tco_z = (otp_data[BMM350_MAG_TCO_Z] & BMM350_LSB_MASK);
+
+	mag_comp.tco[0] = fix_sign(tco_x, 8) / 32.0f;
+	mag_comp.tco[1] = fix_sign(tco_y, 8) / 32.0f;
+	mag_comp.tco[2] = fix_sign(tco_z, 8) / 32.0f;
+
+	tcs_x = (otp_data[BMM350_MAG_TCS_X] & BMM350_MSB_MASK) >> 8;
+	tcs_y = (otp_data[BMM350_MAG_TCS_Y] & BMM350_MSB_MASK) >> 8;
+	tcs_z = (otp_data[BMM350_MAG_TCS_Z] & BMM350_MSB_MASK) >> 8;
+
+	mag_comp.tcs[0] = fix_sign(tcs_x, 8) / 16384.0f;
+	mag_comp.tcs[1] = fix_sign(tcs_y, 8) / 16384.0f;
+	mag_comp.tcs[2] = (fix_sign(tcs_z, 8) / 16384.0f) - BMM350_TCS_CORR_Z;
+
+	mag_comp.t0 = (fix_sign(otp_data[BMM350_MAG_T_0], 16) / 512.0f) + 23.0f;
+
+	cross_x_y = (otp_data[BMM350_CROSS_X_Y] & BMM350_LSB_MASK);
+	cross_y_x = (otp_data[BMM350_CROSS_Y_X] & BMM350_MSB_MASK) >> 8;
+	cross_z_x = (otp_data[BMM350_CROSS_Z_X] & BMM350_LSB_MASK);
+	cross_z_y = (otp_data[BMM350_CROSS_Z_Y] & BMM350_MSB_MASK) >> 8;
+
+	mag_comp.cross_axis.cross_x_y = fix_sign(cross_x_y, 8) / 800.0f;
+	mag_comp.cross_axis.cross_y_x = fix_sign(cross_y_x, 8) / 800.0f;
+	mag_comp.cross_axis.cross_z_x = fix_sign(cross_z_x, 8) / 800.0f;
+	mag_comp.cross_axis.cross_z_y = fix_sign(cross_z_y, 8) / 800.0f;
+
+	return PX4_OK;
+}
+
+void BMM350::compensate_xyzt(float *xyzt)
+{
+	// Apply compensation to temperature */
+
+	xyzt[3] = (1 + mag_comp.sensit_coef_t) * xyzt[3] + mag_comp.offset_coef_t;
+
+	// Apply compensation to x, y, z
+
+	for (int i = 0; i < 3; i++) {
+		xyzt[i] *= 1 + mag_comp.sensit_coef[i];
+		xyzt[i] += mag_comp.offset_coef[i];
+		xyzt[i] += mag_comp.tco[i] * (xyzt[3] - mag_comp.t0);
+		xyzt[i] /= 1 + mag_comp.tcs[i] * (xyzt[3] - mag_comp.t0);
+	}
+
+	float cr_ax_comp_x = (xyzt[0] - mag_comp.cross_axis.cross_x_y * xyzt[1]) /
+			     (1 - mag_comp.cross_axis.cross_y_x * mag_comp.cross_axis.cross_x_y);
+	float cr_ax_comp_y = (xyzt[1] - mag_comp.cross_axis.cross_y_x * xyzt[0]) /
+			     (1 - mag_comp.cross_axis.cross_y_x * mag_comp.cross_axis.cross_x_y);
+	float cr_ax_comp_z = (xyzt[2] +
+			      (xyzt[0] * (mag_comp.cross_axis.cross_y_x * mag_comp.cross_axis.cross_z_y -
+					  mag_comp.cross_axis.cross_z_x) - xyzt[1] *
+			       (mag_comp.cross_axis.cross_z_y - mag_comp.cross_axis.cross_x_y *
+				mag_comp.cross_axis.cross_z_x)) /
+			      (1 - mag_comp.cross_axis.cross_y_x * mag_comp.cross_axis.cross_x_y));
```

Comment:
Can these result in division by zero?

Reply:
I don't think so. The compensation values are very small floats, and the dividers are always close to 1. The equations are copied 1:1 from bosch reference